### ISSUE TITLE
changing the "track" when writing, resets it.

### ISSUE THREAD

**Issue by umlaeute** [pd-externals/midifile#8]
it seems that writing tracks is seriously broken.

most importantly https://github.com/pd-externals/midifile/blob/95354384c269ae2c849fe710c0c13570de5c1656/midifile.c#L1420-L1424 will always be `true` (because `x->track_chunk[x->track].track_data` is nowhere used when writing data), and thus the tempfiles are cleared.

i'm not a frequent user of `[midifile]`, so i might miss something.
otherwise i have a simple fix at hand

afaict, this might be related to one of the subitems of #3

**Comment by danomatika** [comment]
Without looking at the code, I think it's a state mechanism:
1. set track to write to
2. send message
3. repeat for a different track

I would think *only* allowing one track at a time would be severely limiting. Ideally, I could write multiple tracks per tick.

**Comment by umlaeute** [comment]
well, yes; that's what i expect as well.

it seems like it doesn't work like this though.

my fix would be:
```diff
diff --git a/midifile.c b/midifile.c
index 65a427d..bbb84e2 100644
--- a/midifile.c
+++ b/midifile.c
@@ -1417,7 +1417,7 @@ static void midifile_set_track(t_midifile *x, t_floatarg track)
         else
         {
             x->track = track; // possibly update x->header_chunk.chunk_ntrks
-            if (x->track_chunk[x->track].track_data == NULL)
+            if (x->tmpFP[x->track] == NULL)
             {
                 /* this track is being used for the first time */
                 post("this track (%d) is being used for the first time", x->track);
```

(but otoh, i would really like to get rid of these tempfiles altogether (see #7))

**Comment by danomatika** [comment]
Yeah, we don't need temp files. MIDI data is so compact, I can't imagine memory pressure being an issue.

**Comment by danomatika** [comment]
Actually, I think I was mistaking "track" with "channel." Writing multiple channels to a file works fine, as I would expect.

The "track" concept is down to the type of MIDI file (from [wikipedia](https://en.wikipedia.org/wiki/MIDI)):
>A type 0 file contains the entire performance, merged onto a single track, while type 1 files may contain any number of tracks that are performed synchronously. Type 2 files are rarely used[[60]](https://en.wikipedia.org/wiki/MIDI#cite_note-61) and store multiple arrangements, with each arrangement having its own track and intended to be played in sequence.

I believe most software works with type 0 files and the concept of "recording a MIDI song" is often type 0, ala all instruments are channels 1-16 in a single track. I did use seq24 on Linux for some songs and it saved as a track 1 file where the different loops were in different tracks, not just as channels. I had to an extra export mode to output a type 0 file for use with `[midifile]`.

Long story-short, I believe the current implementation is sufficient: realtime playback is mainly outputting events from the type 0 single track, anything with type 1 is probably handled manually by switching tracks to output the events where they can be cached for looping playback, for instance via `[qlist]` or `[text]`.

**Comment by danomatika** [comment]
Another useful line with example MIDI type 0 and type 1 file with the same data: https://www.skytopia.com/project/articles/midi.html
